### PR TITLE
Size Knob

### PR DESCRIPTION
**Pull request by vaibhavrdbhat**

### CODE REVIEW

File: src/test/com/scalyr/api/tests/KnobTest.java [hunk 1]
```diff
@@ -315,43 +315,54 @@
...
     ConfigurationFile bar = factory.getFile("/bar.txt");
 
-    assertEquals(123, (long)(new Knob.Long("number", -1L, bar).get()));
-    assertEquals(23, (long)(new Knob.Long("b", -1L, bar).get()));
-    assertEquals(45 * 1000, (long)(new Knob.Long("kb", -1L, bar).get()));
-    assertEquals(67 * 1024 * 1024, (long)(new Knob.Long("mib", -1L, bar).get()));
-    assertEquals(89L * 1000 * 1000 * 1000, (long)(new Knob.Long("gb", -1L, bar).get()));
-    assertEquals(34L * 1024 * 1024 * 1024 * 1024, (long)(new Knob.Long("tb", -1L, bar).get()));
+    double delta = 0.001; // acceptable error for stuff like converting from -bytes to -ibibytes
+    assertEquals(123,        new Knob.Size("number", -1L, bar).get(), 0              );
```
Comment (jjhart) [inline src/test/com/scalyr/api/tests/KnobTest.java]: DRY - you should write a helper method that does the delta multiplication for you (or just port `TestUtil.assertEquals(double,double,double)`, which supports relative deltas, to this repo).


```
    assertEquals(123,        new Knob.Size("number", -1L, bar).get(), 0    );
    assertEquals(23,         new Knob.Size("b", -1L, bar).getB(),     0    );
    assertEquals(.000023,    new Knob.Size("b", -1L, bar).getMB(),    delta);
```

File: src/test/com/scalyr/api/tests/KnobTest.java [hunk 2]
```diff
@@ -315,43 +315,54 @@
...
 
-    verifyExceptionMessageContains(new Knob.Integer("invalid1", -1, foo)::get, "Can't convert [");
-    verifyExceptionMessageContains(new Knob.Integer("invalid2", -1, foo)::get, "Can't convert [");
-    verifyExceptionMessageContains(new Knob.Integer("invalid3", -1, foo)::get, "Can't convert [");
-    verifyExceptionMessageContains(new Knob.Integer("invalid4", -1, foo)::get, "Can't convert [");
-    verifyExceptionMessageContains(new Knob.Integer("invalid5", -1, foo)::get, "Can't convert [");
+    verifyExceptionMessageContains(new Knob.Size("invalid1", -1L, foo)::get, "Can't convert [");
+    verifyExceptionMessageContains(new Knob.Size("invalid2", -1L, foo)::get, "Can't convert [");
+    verifyExceptionMessageContains(new Knob.Size("invalid3", -1L, foo)::get, "Can't convert [");
+    verifyExceptionMessageContains(new Knob.Size("invalid4", -1L, foo)::get, "Can't convert [");
```
Comment (jjhart) [inline src/test/com/scalyr/api/tests/KnobTest.java]: So verbose!  cf `TestUtils.fails` in the main repo =)

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 3]
```diff
@@ -298,7 +294,6 @@ public synchronized Knob addUpdateListener(Consumer<Knob> updateListener) {
       ensureFileListener();
     }
     updateListeners.add(updateListener);
-
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: Can you add these blank lines back?  I think they help readability, we don't have a policy against them, and so they are complicating the git history for no need.

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 4]
```diff
@@ -113,9 +113,7 @@
    * Specify a default list of configuration files. This will be used for any Knob in which no file list was specified.
    * Existing Knobs are not affected by changes to the default file list.
    */
-  public static void setDefaultFiles(ConfigurationFile[] files) {
-    defaultFiles.set(files);
-  }
+  public static void setDefaultFiles(ConfigurationFile[] files) { defaultFiles.set(files); }
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: Did your editor do this?  I personally _like_ single-line functions like this, but not to the point of encouraging drive-by reformatting of existing files. (also: public repo calls for LCD formatting choices).

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 5]
```diff
@@ -456,55 +434,45 @@ public String(java.lang.String valueKey, java.lang.String defaultValue, Configur
...
    *    micro, micros, microsecond, microseconds, µ, µs
    *    ms, milli, millis, millisecond, milliseconds
    *    s, sec, secs, second, seconds
    *    m, min, mins, minute, minutes
    *    h, hr, hrs, hour, hours
    *    d, day, days
-   *
-   *   - Durations are case insensitive.
-   *   - Spaces are okay when leading, trailing, or in between the amount and the units.
+   *  - Durations are case insensitive, but convention is lowercase
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: `s~is lowercase~is lowercase, to help disambiguate vs SI units on Integer/Long/Size~`

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 6]
```diff
@@ -456,55 +434,45 @@ public String(java.lang.String valueKey, java.lang.String defaultValue, Configur
...
+   *  - Spaces are okay when leading, trailing, or in between the amount and the units.
    *
    * METHODS TO GET VALUE:
-   *
-   *  We provide long-valued accessors that return commonly used units:
+   *  - We provide long-valued accessors that return commonly used units:
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: You need whitespace above lists for our markdown doclet to detect the list (otherwise it formats to run-on text).  I don't think we actually use that doclet in this repo, but it's good practice I think & helps readability.

```
 * METHODS TO GET VALUE:
 * 
 *   - We provide...
```

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 7]
```diff
@@ -456,55 +434,45 @@ public String(java.lang.String valueKey, java.lang.String defaultValue, Configur
...
    *    .nanos(), .micros(), .millis(), .seconds(), .minutes(), .hours(), .days()
-   *
-   *  The standard Knob.get() method is also overridden to return a java.time.Duration object,
-   *  which can be used with its native methods such as .toNanos() to get a Long value.
+   *  - The standard Knob.get() method is also overridden to return a java.time.Duration object,
+   *    which can be used with its native methods such as .toNanos() to get a Long value.
    *
    * EXAMPLE USAGE:
-   *
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: Please restore this whitespace.

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 8]
```diff
@@ -456,55 +434,45 @@ public String(java.lang.String valueKey, java.lang.String defaultValue, Configur
...
    *  // Assume that config file has {myLabel: "1day"}
    *  Knob.Duration myKnob = new Knob.Duration("myLabel", 1L, TimeUnit.SECONDS, paramFile);
    *  long hoursInADay = myKnob.hours(); //Will be 24 hours
    *
    */
   public static class Duration extends Knob {
-
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: And this.

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 9]
```diff
@@ -564,4 +528,61 @@ private long getTimeInNanos(java.lang.Long timeoutInMs, boolean bypassCache) {
       }
     }
   }
+
+  /**
+   * Subclass of Knob which is specialized for sizes, with or without SI.
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: "for sizes" is pretty generic in this javadoc, don't you think?  what's a "size"?

`specialized for byte-denominated sizes (RAM, disk, etc)`

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 10]
```diff
@@ -564,4 +528,61 @@ private long getTimeInNanos(java.lang.Long timeoutInMs, boolean bypassCache) {
...
+   *  // Assume that config file has {myLabel: "1MiB"}
+   *  Knob.Size myKnob = new Knob.Size("myLabel", 1L, paramFile);
+   *  double valueAsKilobytes = myKnob.getKB();
+   */
+  public static class Size extends Knob {
+    public Size(java.lang.String valueKey, java.lang.Long defaultValue, ConfigurationFile ... files) {
+      super(valueKey, defaultValue, files);
+    }
+
+    @Override public java.lang.Double get() { return convertWithSI(super.get()); }
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: This should return `Long` not `Double`, right?

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 11]
```diff
@@ -564,4 +528,61 @@ private long getTimeInNanos(java.lang.Long timeoutInMs, boolean bypassCache) {
...
+    public double getKB()  { return this.get() / 1000D;             } // Kilobyte
+    public double getKiB() { return this.get() / 1024D;             } // Kibibyte
+    public double getMB()  { return this.get() / 1000000D;          } // Megabyte
+    public double getMiB() { return this.get() / 1048576D;          } // Mebibyte
+    public double getGB()  { return this.get() / 1000000000D;       } // Gigabyte
+    public double getGiB() { return this.get() / 1073741824D;       } // Gibibyte
+    public double getTB()  { return this.get() / 1000000000000D;    } // Terabyte
+    public double getTiB() { return this.get() / 1099511627776D;    } // Tebibyte
+    public double getPB()  { return this.get() / 1000000000000000D; } // Petabyte
+    public double getPiB() { return this.get() / 1125899906842624D; } // Pebibyte
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: All of these constants are safer if you derive them, rather than just typing them in.

```
return this.get() / Math.pow(10,9); // Gigabyte
return this.get() / Math.pow(2,30); // Gibibyte
```

The complier will optimize them to constants anyway.

File: src/main/com/scalyr/api/knobs/Knob.java (hunk 10 above)
Comment (vaibhavrdbhat) [reply src/main/com/scalyr/api/knobs/Knob.java]: As the comment in the middle of the function says, all the other `get()`s return doubles for a specific reason, and so to be consistent this one returns a double too. Is that ok?

For the reference, this is the comment:
```
Returns are doubles (for cases such as calling getKB() on '500B', which will yield a decimal).```

File: src/main/com/scalyr/api/knobs/Knob.java (hunk 8 above)
Comment (vaibhavrdbhat) [reply src/main/com/scalyr/api/knobs/Knob.java]: None of the other knobs have a line here, so removed it for consistency.

File: src/main/com/scalyr/api/knobs/Knob.java (hunk 4 above)
Comment (vaibhavrdbhat) [reply src/main/com/scalyr/api/knobs/Knob.java]: Many of the functions like this in Knob.java were already one-line, so since the Java client only gets touched once in a while I figured a cleanup of a couple lines was okay, especially when it was just to achieve consistency.

File: src/main/com/scalyr/api/knobs/Knob.java (hunk 10 above)
Comment (jjhart) [reply src/main/com/scalyr/api/knobs/Knob.java]: Yes, I read that & understood it.  get() by itself should still just return a Long, though.

File: src/main/com/scalyr/api/knobs/Knob.java (hunk 8 above)
Comment (jjhart) [reply src/main/com/scalyr/api/knobs/Knob.java]: Please revert.

File: src/main/com/scalyr/api/knobs/Knob.java (hunk 4 above)
Comment (jjhart) [reply src/main/com/scalyr/api/knobs/Knob.java]: If we want a consistency pass, we'll do it in a different PR.  Drive-by reformatting is not a good habit to get into.

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 12]
```diff
@@ -564,4 +564,69 @@ private long getTimeInNanos(java.lang.Long timeoutInMs, boolean bypassCache) {
...
+   *  - In the config file, define value in the format "[MAGNITUDE] [UNIT]", eg. "2 MiB" or "4ns" or "450 KB".
+   *  - Acceptable units: (Case insensitive, but for convention please format as shown.)
+   *    [no unit == B], B, KB, KiB, MB, MiB, GB, GiB, TB, TiB, PB, PiB
+   *  - Config values must be whole numbers (no decimals)
+   *
+   * METHODS TO GET VALUE:
+   *
+   *  - We provide double-valued accessors as follows:
+   *    .getB(), .getKB(), .getKiB(), .getMB(), .getMiB(), .getGB(), .getGiB(), getTB(), getTiB(), getPB(), getPiB()
+   *  - The standard Knob.get() method will return size in byte amount.
```
Comment (jjhart) [inline src/main/com/scalyr/api/knobs/Knob.java]: `s~size in byte amount~the size in bytes, as a long~`

**Comment by vaibhavrdbhat** [comment]
@jjhart Done!
